data.md

```diff
@@ -97,21 +97,26 @@ These can be used for pattern-matching on the LHS of rules as well (`macro` attr
 
 ```k
     syntax Bool ::= #rangeSInt    ( Int , Int )
-                  | #rangeUInt    ( Int , Int )
+                  | #rangeUInt    ( Int , Int )             [function]
                   | #rangeSFixed  ( Int , Int , Int )
                   | #rangeUFixed  ( Int , Int , Int )
                   | #rangeAddress ( Int )
-                  | #rangeBytes   ( Int , Int )
+                  | #rangeBytes   ( Int , Int )             [function]
  // -------------------------------------------
     rule #rangeSInt    ( 128 ,      X ) => #range ( minSInt128      <= X <= maxSInt128      ) [macro]
     rule #rangeSInt    ( 256 ,      X ) => #range ( minSInt256      <= X <= maxSInt256      ) [macro]
+    rule #rangeUInt    (   8 ,      X ) => #range ( minUInt8        <= X <= maxUInt8        ) [macro]
     rule #rangeUInt    (  48 ,      X ) => #range ( minUInt48       <= X <= maxUInt48       ) [macro]
     rule #rangeUInt    ( 128 ,      X ) => #range ( minUInt128      <= X <= maxUInt128      ) [macro]
     rule #rangeUInt    ( 256 ,      X ) => #range ( minUInt256      <= X <= maxUInt256      ) [macro]
     rule #rangeSFixed  ( 128 , 10 , X ) => #range ( minSFixed128x10 <= X <= maxSFixed128x10 ) [macro]
     rule #rangeUFixed  ( 128 , 10 , X ) => #range ( minUFixed128x10 <= X <= maxUFixed128x10 ) [macro]
     rule #rangeAddress (            X ) => #range ( minUInt160      <= X <= maxUInt160      ) [macro]
     rule #rangeBytes   (  32 ,      X ) => #range ( minUInt256      <= X <= maxUInt256      ) [macro]
+    
+    //General purpose range rules, to define ranges over arbitrary number of bits/bytes. 
+    rule #rangeUInt    (   N ,      X ) => #range ( 0               <= X <=   2 ^Int N -Int 1 )
+    rule #rangeBytes   (   N ,      X ) => #range ( 0               <= X <= 256 ^Int N -Int 1 )
```

Comment:
This doesn't seem to be correct.  Shouldn't it be `2^(N * 8) - 1`?

Reply:
Also, in general, I'm not sure what will happen when there is the overlap between two types of rules. I mean, which rule will be applied for `#rangeUInt(8,X)`?

Reply:
But 2^(N * 8) - 1 = 256^N - 1.

Reply:
Since macro is applied at much earlier stage, it will always apply first. Regular rule will apply only when macro didn't apply.

Reply:
Ah OK. Then please put it in the comment to the right, to save time for others like me.

Reply:
What do you mean by "avoiding two functions"?

Reply:
`#rangeBytes(32,X)` - same as for `#rangeUInt` - macro always applies first.

Reply:
The `#rangeBits` that you propose would be identical to `#rangeUInt`. So why having 2 functions when we can have one?

Reply:
Note that we cannot replace all concrete `#rangeUInt` with the general one. Someone tried but failed. See this: https://github.com/kframework/evm-semantics/pull/246

Reply:
I think the best bet will be:

1. rename the general #rangeUInt to #rangeBits.
2. drop the specific #rangeBytes(32,X) and keep the general one only.

Reply:
The issue in #246 happens when you use general range inside a semantics rule. But inside a spec rule there is an extra pre-processing step that will properly evaluate them, so there won;t be issues. I'm already using this construct in path condition with no issues.

1. Regular `#rangeUInt` has no conflicts with macro `#rangeUInt` so I don't think having different name is a good idea. The only time it would create issues is when someone uses it in a regular rule. Maybe to add a comment to clarify this?

2. #rangeBytes is not used in any regular rules, so yeah macro can be dropped.